src/url_search_params.cpp

```diff
@@ -0,0 +1,60 @@
+#include "ada.h"
+#include "ada/url_search_params.h"
+
+#include <algorithm>
+#include <string>
+#include <string_view>
+#include <vector>
+
+namespace ada {
+
+void url_search_params::remove(const std::string_view key) {
+  params.erase(
+      std::remove_if(params.begin(), params.end(),
+                     [&key](auto &param) { return param.first == key; }),
+      params.end());
+}
+
+void url_search_params::remove(const std::string_view key,
+                               const std::string_view value) {
+  params.erase(std::remove_if(params.begin(), params.end(),
+                              [&key, &value](auto &param) {
+                                return param.first == key &&
+                                       param.second == value;
+                              }),
+               params.end());
+}
+
+void url_search_params::sort() {
+  std::stable_sort(params.begin(), params.end(),
+                   [](const key_value_pair &lhs, const key_value_pair &rhs) {
+                     return lhs.first < rhs.first;
+                   });
+}
+
+void url_search_params::set(const std::string_view key,
+                            const std::string_view value) {
+  params.erase(
+      std::remove_if(params.begin(), params.end(),
+                     [&key](auto &param) { return param.first == key; }),
+      params.end());
+
+  params.emplace_back(key, value);
+}
+
+std::string url_search_params::to_string() {
```

Comment:
This doesn't do percent encoding, does it?

Reply:
https://url.spec.whatwg.org/#concept-urlencoded-serializer

Reply:
Yes we need percent encoding

Reply:
I added a TODO.